### PR TITLE
Implement intersect for ProductIterator (matching CartesianIndices behaviour)

### LINKED ISSUE #39943: Making intersection of Iterators.ProductIterator of AbstractRange a ProductIterator
Follow-up of [this discourse thread](https://discourse.julialang.org/t/intersection-of-productiterator/53357).

Would it be reasonable to have `intersect(::Iterators.ProductIterator{<:NTuple{N,AbstractRange}}, ::Iterators.ProductIterator{<:NTuple{N,AbstractRange}})  where {N}` returning a `ProductIterator` instead of generating all the elements in a `Vector`.

For instance:

```Julia
foo(a::Iterators.ProductIterator{<:NTuple{N,UnitRange}}, b::Iterators.ProductIterator{<:NTuple{N,UnitRange}}) where {N} =
    Iterators.product(map(intersect, a.iterators, b.iterators))
```

Note: I noticed that on v1.6 `intersect` on `CartesianIndices` now has this feature.

### PR DESCRIPTION
**Pull request by sanjibansg**
Fixes #39943

### CODE REVIEW

**Comment by OlivierHnt** [comment]
This pull request chose to implement a specific `intersect` method for `ProductIterator`. In particular, this matches the behaviour of `intersect` for `CartesianIndices`.
This is more general and does not restrict to `AbstractRange`.

I personally like this change. 😃  However shouldn’t the commit title reflect this?
Perhaps something like : “Implement `intersect` for `ProductIterator` (matching `CartesianIndices` behaviour)”

What do you think?

**Comment by kahanikaar** [comment]
> Implement `intersect` for `ProductIterator` (matching `CartesianIndices` behaviour)

Doing the change. PR Title modified.

**Review by OlivierHnt — APPROVED** [review]

**Review by dkarrasch — DISMISSED** [review]
This needs tests, and maybe an announcement in NEWS.

**Comment by OlivierHnt** [comment]
I made two pull requests to your fork @dkarrasch adding tests and an announcement in the NEWS.md file.

**Comment by kahanikaar** [comment]
Tests added. NEWS updated.

File: test/iterators.jl
Comment (dkarrasch) [inline test/iterators.jl:432]:
```suggestion
@test intersect(product(1:3, 4:6), product(2:4, 3:5)) == Iterators.ProductIterator((2:3, 4:5))
```
?

**Comment by kahanikaar** [comment]
@dkarrasch @OlivierHnt Any more changes needed?

**Comment by dkarrasch** [comment]
I think somebody else should take a look here, but I don't know who to ask for a review.

**Comment by OlivierHnt** [comment]
@dkarrasch is there anything we could do to help the review process of this PR ? :)

**Comment by OlivierHnt** [comment]
@kahanikaar could you resolve the conflicts with NEWS.md?
It has been a few weeks since the PR was made so it should be rebased to keep it up-to-date with the master branch of Julia.
Hopefully we'll have some feedback soon.

**Comment by kahanikaar** [comment]
> @kahanikaar could you resolve the conflicts with NEWS.md?
> It has been a few weeks since the PR was made so it should be rebased to keep it up-to-date with the master branch of Julia.
> Hopefully we'll have some feedback soon.

Resolved that.

**Comment by OlivierHnt** [comment]
@dkarrasch can we have some feedback on this? For instance, is this a question about whether or not it is worth doing?

I would like to mention again that this PR simply defines `intersect` method for `ProductIterator` to prevent default behaviour which generates all the elements.

This is completely analogous to the [PR #36643](https://github.com/JuliaLang/julia/pull/36643) which implemented the `intersect` method for `CartesianIndices` and was added to the Julia 1.6 release.

**Comment by dkarrasch** [comment]
This got out of sight, but I guess should be merged (once the merge conflict gets resolved).

**Comment by OlivierHnt** [comment]
@dkarrasch the merge conflict has been resolved and all checks were successful.

**Comment by dkarrasch** [comment]
Thanks for letting me know and for your patience!

**Pull request merged**
